File: python/infinite_sense_core/net.py
import json
import socket
import struct
import threading
import time

from .data import process_trigger_data, process_imu_data, process_gps_data, process_log_data
from .times import precise_sleep
from .ptp import Ptp

class NetManager:
    def __init__(self, target_ip: str, port: int):
        self.target_ip = target_ip
        self.port = port
        self.started = False

        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.setblocking(False)

        curr_time = int(time.time() * 1_000_000)
        self.sock.sendto(struct.pack("<Q", curr_time), (self.target_ip, self.port))

        self.ptp = Ptp()
        self.ptp.set_net_ptr(self.sock, self.target_ip, self.port)

        self.rx_thread = None
        self.tx_thread = None

    def start(self):
        if self.started:
            return
        self.started = True
        self.rx_thread = threading.Thread(target=self.receive, daemon=True)
        self.tx_thread = threading.Thread(target=self.timestamp_synchronization, daemon=True)
        self.rx_thread.start()
        self.tx_thread.start()
        print("[NetManager] Started")

    def stop(self):
        if not self.started:
            return
        self.started = False
        print("[NetManager] Stopped")

    def receive(self):
        while self.started:
            try:
                data, addr = self.sock.recvfrom(65536)
            except BlockingIOError:
                precise_sleep(0.001)
                continue

            try:
                recv_data = data.decode("utf-8")
                if not recv_data:
                    continue
                json_data = json.loads(recv_data)
            except Exception as e:
                print(f"[WARN] JSON parse error: {e}")
                continue
            # print(json_data)
            self.ptp.receive_ptp_data(json_data)
            process_trigger_data(json_data)
            process_imu_data(json_data)
            process_gps_data(json_data)
            process_log_data(json_data)

    def timestamp_synchronization(self):
        while self.started:
            try:
                self.ptp.send_ptp_data()
                precise_sleep(0.01)
            except Exception as e:
                print(f"[ERROR] Timestamp sync error: {e}")


if __name__ == "__main__":
    nm = NetManager("192.168.1.188", 8888)
    nm.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        nm.stop()

File: infinite_sense_core/src/net.cpp
#include "infinite_sense.h"
#include "ptp.h"
#include "data.h"
#include "net.h"

namespace infinite_sense {

NetManager::NetManager(std::string target_ip, const unsigned short port)
    : port_(port), target_ip_(std::move(target_ip)) {
  net_ptr_ = std::make_shared<UDPSocket>();
  const uint64_t curr_time = static_cast<uint64_t>(
      std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now().time_since_epoch())
          .count());
  net_ptr_->sendTo(&curr_time, sizeof(curr_time), target_ip_, port_);
  ptp_ = std::make_unique<Ptp>();
  ptp_->SetNetPtr(net_ptr_, target_ip_, port_);
}

NetManager::~NetManager() {
  Stop();
  if (net_ptr_) {
    net_ptr_->disconnect();
    net_ptr_.reset();
  }
}

void NetManager::Start() {
  if (started_) {
    return;
  }
  started_ = true;
  rx_thread_ = std::thread(&NetManager::Receive, this);
  tx_thread_ = std::thread(&NetManager::TimeStampSynchronization, this);
  LOG(INFO) << "Net manager started";
}

void NetManager::Stop() {
  if (!started_) {
    return;
  }
  started_ = false;
  if (rx_thread_.joinable()) {
    rx_thread_.join();
  }
  if (tx_thread_.joinable()) {
    tx_thread_.join();
  }
  LOG(INFO) << "Net manager stopped";
}

void NetManager::Receive() const {
  constexpr size_t k_buffer_size = 65536;
  std::array<uint8_t, k_buffer_size> buffer{};
  std::string source_address;
  unsigned short source_port = 0;

  while (started_) {
    const int size = net_ptr_->recvFrom(buffer.data(), k_buffer_size, source_address, source_port);
    if (size <= 0) {
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
      continue;
    }

    try {
      std::string recv_data(reinterpret_cast<char*>(buffer.data()), size);
      if (recv_data.empty()) {
        continue;
      }
      auto json_data = nlohmann::json::parse(recv_data, nullptr, false);
      if (json_data.is_discarded()) {
        LOG(WARNING) << "Received malformed JSON";
        continue;
      }
      ptp_->ReceivePtpData(json_data);
      ProcessTriggerData(json_data);
      ProcessIMUData(json_data);
      ProcessGPSData(json_data);
      ProcessLOGData(json_data);

    } catch (const std::exception& e) {
      LOG(ERROR) << "Exception during JSON handling: " << e.what();
    }
  }
}

void NetManager::TimeStampSynchronization() const {
  while (started_) {
    try {
      ptp_->SendPtpData();
      std::this_thread::sleep_for(std::chrono::milliseconds(10));
    } catch (const std::exception& e) {
      LOG(ERROR) << "Timestamp sync error: " << e.what();
    }
  }
}

}  // namespace infinite_sense
